File: lua/high-scores/high-scores.lua
-- A closure-based solution.
-- Colon notation will pass a "self" parameter to the functions:
-- we can ignore it.

return function(values)
  return {
    scores = function()
      return values
    end,

    latest = function()
      return values[#values]
    end,

    personal_best = function()
      return math.max(table.unpack(values))
    end,

    personal_top_three = function()
      local clone = { table.unpack(values) }
      table.sort(clone, function(a, b) return a > b end) -- reverse sort
      top3 = { table.unpack(clone, 1, 3) }
      return top3
    end
  }
end

File: php/high-scores/HighScores.php
<?php

declare(strict_types=1);

class HighScores
{
    public array $scores;
    public int $latest;
    public int $personalBest;
    public array $personalTopThree;

    public function __construct(array $scores)
    {
        $this->scores = $scores;
        $this->latest = end($scores);
        $this->personalBest = max(PHP_INT_MIN, ...$scores);
        rsort($scores);
        $this->personalTopThree = array_slice($scores, 0, 3);
    }
}
